docs/blog/november-contributions-showcase.md

```diff
@@ -0,0 +1,111 @@
+---
+title: November contributions showcase
+author:
+  - name: James Garbutt
+sidebar: false
+date: 2024-12-03
+head:
+  - - meta
+    - property: og:type
+      content: website
+  - - meta
+    - property: og:title
+      content: November contributions showcase
+  - - meta
+    - property: og:url
+      content: https://e18e.dev/blog/november-contributions-showcase
+  - - meta
+    - property: og:description
+      content: A showcase of e18e-focused contributions from November
+  - - meta
+    - property: og:image
+      content: https://e18e.dev/og/november-contributions-showcase.png
+---
+
+# November contributions
+
+_December 3, 2024_
+
+Another month, another great list of performance gains across the ecosystem. This month, we've also seen a huge effort to improving visibility of these things so more people can get involved. Let's take a look!
+
+## Libraries
+
+### kasi
+
+A set of utilities we've seen quite often in the wild are case transforms. For example, transforming to and from camel case (e.g. from `foo-bar` to `fooBar`).
+
+Often, this has been achieved by using a micro-utility dependency or by repeating a very similar function to everyone else who needed this logic.
+
+This is a really good example of something which is too big to be copied everywhere, but too small to justify its own package.
+
+So what do we do in these cases? We group the common case transforms and use _that_ as a library.
+
+That is exactly what Fabio has done with [kasi](https://github.com/fabiospampinato/kasi)!
+
+kasi is a great collection of functions for transforming between various cases (pascal, camel, snake, etc). Instead of pulling in many micro-packages, or repeating code, we can now use this.
+
+Great work as always, Fabio! :pray:
+
+### unicode-segmenter
+
+As part of the bluesky contributions mentioned later in this post, some of the community noticed a large chunk of the JS bundle was taken up by [graphemer](https://github.com/flmnt/graphemer).
+
+This library is used to split or count the number of grapheme clusters in a string (i.e. the visual symbols we would perceive as individual characters). It does a good job at this but comes in at a whopping 812KB!
+
+Fortunately, [@hyeseong.kim](https://bsky.app/profile/hyeseong.kim) came along and built [unicode-segmenter](https://github.com/cometkim/unicode-segmenter) :tada: This library achieves the same functionality at a fraction of the size (250KB).
```

Comment:
How do you get this number (250KB) ? The same grapheme functionality is only takes 17313 bytes even without minification in the latest version (v0.11.2)

Reply:
You can check it from [bundlejs](https://bundlejs.com/?q=unicode-segmenter%2Fgrapheme&treeshake=%5B*%5D) (![](https://deno.bundlejs.com/badge?q=unicode-segmenter/grapheme&treeshake=[*]))
More detail stats is on README: https://github.com/cometkim/unicode-segmenter?tab=readme-ov-file#js-bundle-stats

Reply:
interesting, both [pkg-size](https://pkg-size.dev/unicode-segmenter) and [npm](https://www.npmjs.com/package/unicode-segmenter) say 250KB

maybe because I'm looking at overall package size rather than just the grapheme entry point?

Reply:
Oh ok that means NPM package size, not bundle size? That makes sense, but 250KB seems to large still, let me check a quick

Reply:
The package is bloated mostly by commonjs entries and its sourcemap :sweat_smile: 

It is not the actual bundle size anyway, the exact size of the graphemer or Intl alternative is much smaller due to the tree-shaking.

<details>

<summary> npm pack </summary>

```
npm notice
npm notice 📦  unicode-segmenter@0.11.2
npm notice Tarball Contents
npm notice 2.4kB _emoji_data.cjs
npm notice 2.1kB _emoji_data.cjs.map
npm notice 618B _emoji_data.d.ts
npm notice 1.7kB _emoji_data.js
npm notice 11.7kB _general_data.cjs
npm notice 11.5kB _general_data.cjs.map
npm notice 711B _general_data.d.ts
npm notice 11.1kB _general_data.js
npm notice 13.0kB _grapheme_data.cjs
npm notice 15.0kB _grapheme_data.cjs.map
npm notice 3.0kB _grapheme_data.d.ts
npm notice 14.0kB _grapheme_data.js
npm notice 1.3kB _incb_data.cjs
npm notice 838B _incb_data.cjs.map
npm notice 311B _incb_data.d.ts
npm notice 589B _incb_data.js
npm notice 1.1kB LICENSE
npm notice 13.6kB README.md
npm notice 2.0kB core.cjs
npm notice 3.9kB core.cjs.map
npm notice 3.7kB core.d.ts
npm notice 2.8kB core.js
npm notice 1.5kB emoji.cjs
npm notice 1.3kB emoji.cjs.map
npm notice 647B emoji.d.ts
npm notice 937B emoji.js
npm notice 1.6kB general.cjs
npm notice 1.5kB general.cjs.map
npm notice 666B general.d.ts
npm notice 979B general.js
npm notice 5.9kB grapheme.cjs
npm notice 13.3kB grapheme.cjs.map
npm notice 2.2kB grapheme.d.ts
npm notice 9.6kB grapheme.js
npm notice 1.2kB index.cjs
npm notice 339B index.cjs.map
npm notice 117B index.d.ts
npm notice 131B index.js
npm notice 3.4kB intl-adapter.cjs
npm notice 4.5kB intl-adapter.cjs.map
npm notice 2.1kB intl-adapter.d.ts
npm notice 3.1kB intl-adapter.js
npm notice 402B intl-polyfill.cjs
npm notice 707B intl-polyfill.cjs.map
npm notice 11B intl-polyfill.d.ts
npm notice 342B intl-polyfill.js
npm notice 1.1kB licenses/unicode-segmentation_MIT.txt
npm notice 4.3kB package.json
npm notice 1.7kB src/_emoji_data.js
npm notice 11.1kB src/_general_data.js
npm notice 14.0kB src/_grapheme_data.js
npm notice 589B src/_incb_data.js
npm notice 2.8kB src/core.js
npm notice 937B src/emoji.js
npm notice 979B src/general.js
npm notice 9.6kB src/grapheme.js
npm notice 131B src/index.js
npm notice 3.1kB src/intl-adapter.js
npm notice 342B src/intl-polyfill.js
npm notice 3.0kB src/utils.js
npm notice 2.4kB utils.cjs
npm notice 4.3kB utils.cjs.map
npm notice 2.1kB utils.d.ts
npm notice 3.0kB utils.js
npm notice Tarball Details
npm notice name: unicode-segmenter
npm notice version: 0.11.2
npm notice filename: unicode-segmenter-0.11.2.tgz
npm notice package size: 58.3 kB
npm notice unpacked size: 239.0 kB
npm notice shasum: f15eca105936dd7f48075b9b7c0d915066d58f6a
npm notice integrity: sha512-ssdWyVdO9kfmt[...]R4Kdidgdw1ThA==
npm notice total files: 64
npm notice
unicode-segmenter-0.11.2.tgz
```

</summary>

Reply:
ESM entries seems to be duplicated. It uses the source as is without any building, so there is no reason to include `*.js` once more in the root. Good catch

Reply:
in the blog posts so far, we've been using _install size_, so we should stick with that here too

it sounds like you will be able to make it smaller though. it is still pretty small compared to grapheme anyway 

I might still mention bundle size just to be clear though

Reply:
If I build CommonJS entries in a more compact format, they don't even need a source map as they are fully readable. Thus, the total package size can be reduced to 135KB !

![image](https://github.com/user-attachments/assets/10f94464-6a02-4bee-9b4e-68f84691cdd1)

Reply:
Of course it will be even much better if it uses `require(ESM)` :joy:

Reply:
https://github.com/cometkim/unicode-segmenter/releases/tag/unicode-segmenter%400.11.3